File: sdk/go/agent/session_transport.go
package agent

import (
	"fmt"
	"sort"
	"strings"
)

var SupportedSessionTransports = map[string][]string{
	"openai":     {"webrtc", "websocket"},
	"openrouter": {"audio_turns"},
}

type SessionTransportCapability struct {
	Provider  string
	Transport string
}

type SessionTransportError struct {
	Provider  string
	Transport string
	Supported []string
}

func (e *SessionTransportError) Error() string {
	return fmt.Sprintf(
		"unsupported session transport %q for provider %q. Supported transports: %s. AgentField does not infer or switch providers; set provider and transport explicitly.",
		e.Transport,
		e.Provider,
		strings.Join(e.Supported, ", "),
	)
}

func NormalizeSessionTransportValue(value string) string {
	return strings.ReplaceAll(strings.ToLower(strings.TrimSpace(value)), "-", "_")
}

func ValidateSessionTransport(provider string, transport string) (SessionTransportCapability, error) {
	normalizedProvider := NormalizeSessionTransportValue(provider)
	normalizedTransport := NormalizeSessionTransportValue(transport)

	if normalizedProvider == "" {
		return SessionTransportCapability{}, fmt.Errorf("session provider is required; AgentField does not infer providers")
	}
	if normalizedTransport == "" {
		return SessionTransportCapability{}, fmt.Errorf("session transport is required; AgentField does not infer transports")
	}

	supported, ok := SupportedSessionTransports[normalizedProvider]
	if !ok {
		known := make([]string, 0, len(SupportedSessionTransports))
		for knownProvider := range SupportedSessionTransports {
			known = append(known, knownProvider)
		}
		sort.Strings(known)
		return SessionTransportCapability{}, fmt.Errorf(
			"unknown session provider %q. Known providers: %s. Register provider capabilities before using a custom session provider",
			provider,
			strings.Join(known, ", "),
		)
	}

	for _, candidate := range supported {
		if candidate == normalizedTransport {
			return SessionTransportCapability{
				Provider:  normalizedProvider,
				Transport: normalizedTransport,
			}, nil
		}
	}

	return SessionTransportCapability{}, &SessionTransportError{
		Provider:  normalizedProvider,
		Transport: normalizedTransport,
		Supported: append([]string(nil), supported...),
	}
}

File: sdk/typescript/src/sessionTransport.ts
export const SUPPORTED_SESSION_TRANSPORTS = {
  openai: ['webrtc', 'websocket'],
  openrouter: ['audio_turns']
} as const;

export type SessionProvider = keyof typeof SUPPORTED_SESSION_TRANSPORTS;
export type SessionTransport = (typeof SUPPORTED_SESSION_TRANSPORTS)[SessionProvider][number];

export interface SessionTransportCapability {
  provider: SessionProvider;
  transport: SessionTransport;
}

export class SessionTransportError extends Error {
  readonly provider: string;
  readonly transport: string;
  readonly supported: readonly string[];

  constructor(provider: string, transport: string, supported: readonly string[]) {
    const supportedDisplay = supported.length > 0 ? supported.join(', ') : 'none';
    super(
      `Unsupported session transport '${transport}' for provider '${provider}'. ` +
        `Supported transports: ${supportedDisplay}. AgentField does not infer ` +
        'or switch providers; set provider and transport explicitly.'
    );
    this.name = 'SessionTransportError';
    this.provider = provider;
    this.transport = transport;
    this.supported = supported;
  }
}

export function normalizeSessionTransportValue(value: string): string {
  return value.trim().toLowerCase().replace(/-/g, '_');
}

export function validateSessionTransport(provider: string, transport: string): SessionTransportCapability {
  const normalizedProvider = normalizeSessionTransportValue(provider);
  const normalizedTransport = normalizeSessionTransportValue(transport);

  if (!normalizedProvider) {
    throw new Error('Session provider is required; AgentField does not infer providers.');
  }
  if (!normalizedTransport) {
    throw new Error('Session transport is required; AgentField does not infer transports.');
  }

  const supported = SUPPORTED_SESSION_TRANSPORTS[normalizedProvider as SessionProvider];
  if (!supported) {
    const known = Object.keys(SUPPORTED_SESSION_TRANSPORTS).sort().join(', ');
    throw new Error(
      `Unknown session provider '${provider}'. Known providers: ${known}. ` +
        'Register provider capabilities before using a custom session provider.'
    );
  }

  if (!supported.includes(normalizedTransport as never)) {
    throw new SessionTransportError(normalizedProvider, normalizedTransport, supported);
  }

  return {
    provider: normalizedProvider as SessionProvider,
    transport: normalizedTransport as SessionTransport
  };
}
